[ownership] update sroa to support ossa

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This patch updates `SROA` to support ossa. I've successfully compiled the stdlib after removing ownership elimination but, like the other similar patches, I think the best way to test this will be to wait until `addPerfEarlyModulePassPipeline` supports ownership. 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

zoecarver commented:
@swift-ci please test.

zoecarver commented on test/SILOptimizer/sroa_ossa.sil:
```diff
@@ -0,0 +1,440 @@
+// RUN: %target-sil-opt -enable-sil-verify-all -sroa %s | %FileCheck %s
```
I'd like to add some non-trivial tests. I might do that as its own patch though.

gottesmm commented:
I am not sure if this is the right patch. I don't know if we want to create borrow scopes everywhere. In certain cases we may want to use destructure. Not sure. Would need to think about it more.

zoecarver commented:
If I understand correctly, using struct extracts is helpful because it allows better results from other optimizations. As you said, though, it may not _always_ be best. I'm happy to update this patch to use `destructure_struct` or whatever you think is best. Just let me know.

atrick commented:
@zoecarver @gottesmm My current position is that it's always best to avoid borrow scopes except in two cases involving address types:
1. to guard uses of an interior pointer
2. to eliminate a copies to/from addressable memory

Using borrows around normal SIL values prevents analysis of their individual lifetimes. I strongly prefer destructure--the only concern is SIL bloat. I could prove to be wrong about all this though.

zoecarver commented:
@atrick I see the concern. That makes a lot of sense. I'll update this to use destructure, then.

zoecarver commented:
Should I use destructure in both ossa and normal functions? Either way, I think I'll put together a separate patch to update this pass to use destructure so that we can see what the performance implications are on their own.

atrick commented:
@zoecarver destructure only exists in OSSA. I would like OSSA/non-OSSA instructions to be prominently documented in SIL.rst (begin_borrow doesn't even exist). SIL instruction documentation is basically spread across SIL.rst (only the format and semantics of the instruction), SILInstruction.h (not much more than SIL.rst), SILBuilder, where OSSA/non-OSSA is verified, and SILVerifier.cpp, which is the only true definition of an instruction's "IR contract"

zoecarver commented:
@atrick very good to know. I've never come across destructure before so I didn't know much about it. It's actually very helpful for mem2reg as well. I think I'll just update this patch, in that case. We should probably try to add support for it in more of the visitors that expect `struct_extract`, `ref_element_addr`, etc. (that's probably obvious, though).

zoecarver committed:
[ownership] Update SROA to support OSSA.

Use destructure to chop up allocations. Other than that, just add
correct load/store qualifiers.

zoecarver commented:
@atrick @gottesmm sorry for the delay updating this patch. I forgot about many of these ownership patches while working on other PRs. I've updated this to use destructure instead of borrow scopes.

CodaFi commented:
@swift-ci test

zoecarver commented:
Uninitialized pointer 😕 I don't know why I always forget but I've actually been bitten by that a few times in the past week. Maybe we should turn on the clang warning/error.

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test

zoecarver commented:
@swift-ci please test windows platform

gottesmm commented:
Before this lands I am going to want to look at this.

zoecarver commented:
Sounds good. No rush. I just wanted to make sure it was green when you're ready.

gottesmm commented on lib/SILOptimizer/Transforms/SILSROA.cpp:
```diff
@@ -249,9 +249,14 @@ void SROAMemoryUseAnalyzer::chopUpAlloca(std::vector<AllocStackInst *> &Worklist
+    for (auto *NewAI : NewAllocations) {
+      LoadOwnershipQualifier loadQual = LoadOwnershipQualifier::Unqualified;
+      if (LI->getFunction()->hasOwnership())
+        loadQual = NewAI->getType().isTrivial(*LI->getFunction())
```
This is not needed. You should be using emitLoadValueOperation. That should work in all cases.

gottesmm commented on lib/SILOptimizer/Transforms/SILSROA.cpp:
```diff
@@ -260,12 +265,32 @@ void SROAMemoryUseAnalyzer::chopUpAlloca(std::vector<AllocStackInst *> &Worklist
+    // In ownership mode, create a destructure of the source and use that to
+    // initialize the aggregate.
+    MultipleValueInstruction *destructured = nullptr;
+    if (SI->getFunction()->hasOwnership()) {
```
Rather than do it this way, there are helpers for destructures that work regardless of ossa or non-ossa. You want to use:

https://github.com/apple/swift/blob/7d2f9440866c5360cc4929f003850058ecfc98bc/include/swift/SIL/SILBuilder.h#L1470

gottesmm commented on lib/SILOptimizer/Transforms/SILSROA.cpp:
```diff
@@ -260,12 +265,32 @@ void SROAMemoryUseAnalyzer::chopUpAlloca(std::vector<AllocStackInst *> &Worklist
+    }
+    for (unsigned eltNo : indices(NewAllocations)) {
+      if (destructured)
+        builder.createStore(SI->getLoc(), destructured->getResult(eltNo),
```
This should be an emitStoreValueOperation. I think there is a bunch of conditionality that is not needed if you used emitDestructureValueOperation.

zoecarver commented:
@gottesmm thanks for the review! Updated based on your comments.

zoecarver commented on test/SILOptimizer/sroa.sil:
```diff
@@ -23,10 +23,10 @@ sil @use_int32 : $@convention(thin) (Builtin.Int32) -> ()
 // CHECK:   [[VAR_2:%[0-9]+]] = alloc_stack $Builtin.Int32
```
I've only moved lines up and down. I haven't changed or removed anything (beyond that).

gottesmm commented:
@zoecarver I think the createAggProjection method is now dead.

zoecarver commented:
@gottesmm you're right. Updated :)

zoecarver commented:
@gottesmm friendly ping.

zoecarver commented:
Closing in favor of #33260.
